lib/audit-report.js

```diff
@@ -110,112 +100,143 @@ class AuditReport extends Map {
 
   async run () {
     this.report = await this[_getReport]()
-    if (this.report) {
+    if (this.report)
       await this[_init]()
-      await this[_processDeps]()
-    }
     return this
   }
 
+  isVulnerable (node) {
+    const vuln = this.get(node.package.name)
+    return !!(vuln && vuln.isVulnerable(node))
+  }
+
   async [_init] () {
     process.emit('time', 'auditReport:init')
 
     const promises = []
     for (const [name, advisories] of Object.entries(this.report)) {
       for (const advisory of advisories) {
-        const { vulnerable_versions: range } = advisory
-        promises.push(this[_addVulnerability](name, range, advisory))
+        //const { vulnerable_versions: range } = advisory
+        promises.push(this.calculator.calculate(name, advisory))
       }
     }
-    await Promise.all(promises)
 
-    process.emit('timeEnd', 'auditReport:init')
-  }
+    // now the advisories are calculated with a set of versions
+    // and the packument.  turn them into our style of vuln objects
+    // which also have the affected nodes, and also create entries
+    // for all the metavulns that we find from dependents.
+    const advisories = new Set(await Promise.all(promises))
+    const seen = new Set()
+    for (const advisory of advisories) {
+      const { name, range } = advisory
+
+      // don't flag the exact same name/range more than once
+      // adding multiple advisories with the same range is fine, but no
+      // need to search for nodes we already would have added.
+      const k = `${name}@${range}`
+      if (seen.has(k)) {
+        continue
+      }
+      seen.add(k)
 
-  // for each node P
-  //  for each vulnerable dep Q
-  //    pickManifest(Q, P's dep on Q, {avoid})
-  //    if resulting version is vunlerable, then P@version is vulnerable
-  //      find all versions of P depending on unsafe Q
-  async [_processDeps] () {
-    process.emit('time', 'auditReport:process')
-    for (const p of this[_vulnDependents]) {
-      await this[_processDependent](p)
-    }
-    process.emit('timeEnd', 'auditReport:process')
-  }
+      const vuln = this.get(name) || new Vuln({ name, advisory })
+      if (this.has(name))
+        vuln.addAdvisory(advisory)
+      super.set(name, vuln)
```

Comment:
is there a reason for using `this` (`this.get`, `this.has`) elsewhere but then this one is using `super` (`super.set`) ?

Reply:
`this.set` throws an error, because you're not supposed to use `report.set()` directly.  Calling `super.set()` calls the method from the Map class (which this extends).

Reply:
right! I missed the `set()` method! 😄